oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/OAuth2AuthorizationCodeGrantFilter.java

```diff
@@ -166,7 +197,8 @@ private void processAuthorizationResponse(HttpServletRequest request, HttpServle
 
 		MultiValueMap<String, String> params = OAuth2AuthorizationResponseUtils.toMultiMap(request.getParameterMap());
 		String redirectUri = UriComponentsBuilder.fromHttpUrl(UrlUtils.buildFullRequestUrl(request))
-				.replaceQuery(null)
+				.replaceQueryParam(OAuth2ParameterNames.CODE)
+				.replaceQueryParam(OAuth2ParameterNames.STATE)
 				.build()
 				.toUriString();
 		OAuth2AuthorizationResponse authorizationResponse = OAuth2AuthorizationResponseUtils.convert(params, redirectUri);
```

Comment:
If the authorization server adds additional parameters to the authorization response, than the `OAuth2AuthorizationResponse.redirectUri` will contain the additional parameters and result in a failed validation check in `OAuth2AuthorizationCodeAuthenticationProvider` and the call to `OAuth2AuthorizationExchangeValidator.validate()` when it compares the `redirectUri` between the authorization request and response. This revealed that we need an integration test for this scenario. Can you please add an integration test and fix so it passes. Take a look at `OAuth2LoginTests` or `OAuth2LoginConfigurerTests` for an example for the integration test.

Reply:
ah, true.
So, now the question is how you want to share the validation logic.
My original PR introduced `OAuth2AuthorizationExchangeUtils` class to share the validation logic between `OAuth2AuthorizationCodeGrantFilter` and `OAuth2AuthorizationExchangeValidator`. (also `OidcAuthorizationCodeAuthenticationProvider`). ([link](https://github.com/ttddyy/spring-security/commit/52dcd13e31439491de9c93086c9e7a9b74886d07#diff-da51724beb161d9b530f7494b5b78daeR36) bit outdated)

This is because `OAuth2AuthorizationResponseUtils` seems good place to put the validation logic but it is package scoped class.
Since `~CodeGrantFilter` and `~ExchangeValidator` are in different package, we need a way to share this validation logic. So, if you don't oppose, I'm going to put back `OAuth2AuthorizationExchangeUtils` to share the redirect url validation logic.

Reply:
> So, now the question is how you want to share the validation logic.
My original PR introduced `OAuth2AuthorizationExchangeUtils` class to share the validation logic between `OAuth2AuthorizationCodeGrantFilter` and `OAuth2AuthorizationExchangeValidator`

We do not want to share the logic between `OAuth2AuthorizationCodeGrantFilter` and `OAuth2AuthorizationExchangeValidator`. Reason being is that `OAuth2AuthorizationCodeGrantFilter` is in the `web` sub-package and the `OAuth2AuthorizationExchangeValidator` is in the `authentication` sub-package, which is NOT a web component.

The validation check in `OAuth2AuthorizationExchangeValidator` can simply be changed to:

```java
if (!authorizationResponse.getRedirectUri().startsWith(authorizationRequest.getRedirectUri()))
```

It's safe to use `startsWith`, since we know that any additional query parameters will be **appended** to the registered `redirect-uri`

> The redirection endpoint URI MUST be an absolute URI as defined by
[RFC3986] Section 4.3. The endpoint URI MAY include an
"application/x-www-form-urlencoded" formatted (per Appendix B) query
component ([RFC3986] Section 3.4), which MUST be retained when **adding
additional query parameters**. The endpoint URI MUST NOT include a
fragment component.